### PR TITLE
[gpt-oss] Update triton_kernels mxfp4 backend to be compatible with triton_kernels upstream

### PR DESCRIPTION
**Pull request by zyongye**
Recently, triton_kernels introduced routing function [refactoring](https://github.com/triton-lang/triton/pull/8375). We need to incorporate it into vllm. 

Partial address #26582

Accuracy bench
openai/gpt-oss-120b, TP=2, GPQA:
| Reasoning effort | Score |
|:--------|:-----:|
| Low | 65.84  | 
| Medium  | 72.2  |

### CODE REVIEW

File: vllm/model_executor/layers/fused_moe/gpt_oss_triton_kernels_moe.py [hunk 1]
```diff
@@ -71,6 +83,50 @@ def pack_bitmatrix(
         tl.store(bitmatrix_ptrs, y, mask=offsets_m[:, None] < n_rows)
 
 
+def routing_from_bitmatrix(bitmatrix, expt_scal, expt_indx, n_expts_tot, n_expts_act):
```
Comment (bnellnm) [inline vllm/model_executor/layers/fused_moe/gpt_oss_triton_kernels_moe.py]: Could you add type annotations to this function?

File: vllm/model_executor/layers/fused_moe/gpt_oss_triton_kernels_moe.py [hunk 2]
```diff
@@ -71,6 +83,50 @@ def pack_bitmatrix(
...
+        n_expts_tot,
+        n_expts_act,
+        ragged_batch_metadata,
+    )
+    gather_idx = GatherIndx(combine_indx, dispatch_index)
+    scatter_idx = ScatterIndx(dispatch_index, combine_indx)
+    return routing_data, gather_idx, scatter_idx
+
+
+def routing(logits, n_expts_act, sm_first=False, expt_indx=None, n_rows=None):
```
Comment (bnellnm) [inline vllm/model_executor/layers/fused_moe/gpt_oss_triton_kernels_moe.py]: Ditto

**Review by bnellnm — APPROVED** [review]
LGTM

File: vllm/model_executor/layers/fused_moe/gpt_oss_triton_kernels_moe.py (hunk 2 above)
Comment (zyongye) [reply vllm/model_executor/layers/fused_moe/gpt_oss_triton_kernels_moe.py]: All comments addressed.

**Comment by eppaneamd** [comment]
Hello @zyongye! Just FYI (looking into this): 

```
(APIServer pid=7223) INFO 10-17 04:46:12 [model.py:658] Resolved architecture: Qwen3VLMoeForConditionalGeneration
(APIServer pid=7223) INFO 10-17 04:46:12 [model.py:1745] Using max model len 262144
(APIServer pid=7223) ERROR 10-17 04:46:12 [gpt_oss_triton_kernels_moe.py:40] Failed to import Triton kernels. Please make sure your triton version is compatible. Error: No module named 'triton_kernels.tensor'
```

**Comment by zyongye** [comment]
@eppaneamd Is the import error still present for models that don't require `triton_kernels`?

**Comment by mgoin** [comment]
I agree with Varun, is there a reason not to target the triton release that matches torch 2.9 and add triton_kernels as a proper dep at that point?

**Comment by zyongye** [comment]
> I agree with Varun, is there a reason not to target the triton release that matches torch 2.9 and add triton_kernels as a proper dep at that point?

No particular reason. We can defer merging this after a new torch release. But we definitely need trying to make triton_kernels as a dep and pin to the triton release commit.

**Comment by zyongye** [comment]
> > @eppaneamd Is the import error still present for models that don't require `triton_kernels`?
> 
> That's a good question, unfortunately not sure. Haven't been testing that extensively, but briefly with `Qwen3VLMoeForConditionalGeneration`. I have been looking for possible solutions/workarounds for getting it work, as I am encountering an import error with some of these (and so far has been difficult to find a commit where all of these work as vllm currently expects)
> 
> ```
> triton_kernels.routing.routing_from_bitmatrix
> triton.compiler.compiler.triton_key
> triton_kernels.tensor
> triton_kernels.tensor.SparseMatrix
> ```

@eppaneamd I think you need to update `triton` to main and use corresponding `triton_kernels` in that commit.

**Comment by eppaneamd** [comment]
> > > @eppaneamd Is the import error still present for models that don't require `triton_kernels`?
> > 
> > 
> > That's a good question, unfortunately not sure. Haven't been testing that extensively, but briefly with `Qwen3VLMoeForConditionalGeneration`. I have been looking for possible solutions/workarounds for getting it work, as I am encountering an import error with some of these (and so far has been difficult to find a commit where all of these work as vllm currently expects)
> > ```
> > triton_kernels.routing.routing_from_bitmatrix
> > triton.compiler.compiler.triton_key
> > triton_kernels.tensor
> > triton_kernels.tensor.SparseMatrix
> > ```
> 
> @eppaneamd I think you need to update `triton` to main and use corresponding `triton_kernels` in that commit.

That yields triton_key import error unfortunately

**Comment by eppaneamd** [comment]
> Context: At the moment, to get `triton_kernels` working with vllm we need to install `triton_kernels` explicitly. For DP, we just have to build from source as it requires torch nightly - more information here #26582
> 
> Discussed with @zyongye over DM: This PR introduces changes so vllm works with triton_kernels from triton `main`. I believe we shouldn't target triton `main` as it is too volatile - IIUC, `torch2.9` ships with Triton branch `release/3.5.x` and we should also target `triton_kernels` in `release/3.5.x` so it works for when vllm starts supporting `torch2.9`. For now, the workarounds mentioned in issue #26582 should be sufficient.
> 
> fwiw, I tried the following to see if `triton_kernels` works with vllm for gpt-oss,
> 
> ```
>  # build vllm from source using torch2.9 
>  uv pip install torch==2.9 
>  uv pip install torchaudio 
>  uv pip install torchvision 
>  git clone https://github.com/vllm-project/vllm.git
>  cd vllm/
>  python use_existing_torch.py
>  uv pip install requirements/common.txt 
>  uv pip install -r requirements/common.txt 
>  uv pip install -r requirements/cuda.txt 
>  uv pip install -r requirements/build.txt 
>  uv pip install -vvv  -e .  --no-build-isolation 
>  
> # Installed triton_kernels like,
> git clone https://github.com/triton-lang/triton.git 
> cd triton
> git checkout release/3.5.x
> cd ../
> uv pip install triton/python/triton_kernels --no-deps 
> ```
> 
> I see that it works for both TP and DP on vllm `main` - when we switch to torch2.9, we could add `triton_kernels` as a dependency.
> 
> @zou3519 @ProExpertProg @mgoin - Is it correct to target Triton branch `release/3.5.x` ? @zyongye I believe I've captured our conversation correctly. Please feel free to correct me.

@varun-sundar-rabindranath do you have any estimate when vllm starts to officially support `torch2.9`?
